consensus/integration_test.go

```diff
@@ -0,0 +1,247 @@
+package consensus
+
+import (
+	"testing"
+	"time"
+
+	"github.com/deso-protocol/core/collections"
+	"github.com/holiman/uint256"
+	"github.com/stretchr/testify/require"
+)
+
+// This integration test simulates a network of 4 nodes with equal stake. It tests the
+// a scenario in which a super-majority of stake is always online. Nodes sporadically
+// go offline and come back online. The network should continue to produce blocks
+// as long as a super-majority of stake is online.
+func TestNetworkWithOfflineValidators(t *testing.T) {
+	// Create 4 nodes with equal stake. The network has super-majority of stake online
+	// as long as 3 out of 4 nodes are online.
+	node1 := newValidatorNode(uint256.NewInt().SetUint64(50), true)  // block proposer
+	node2 := newValidatorNode(uint256.NewInt().SetUint64(50), false) // validator
+	node3 := newValidatorNode(uint256.NewInt().SetUint64(50), false) // validator
+	node4 := newValidatorNode(uint256.NewInt().SetUint64(50), false) // validator
+
+	allNodes := []*validatorNode{node1, node2, node3, node4}
+
+	// Create the genesis block signed by all four nodes.
+	genesisBlock := createDummyBlockWithVoteQC(createDummyBlockHash(), 2, 2, allNodes)
```

Comment:
Maybe dumb question: Why does the genesis block start with view=2 block height=2 rather than (0,0)? I assume it's because some checks fail if you do the latter? Leave a comment to explain either way.

Reply:
The event loop rejects blocks with view=0 because it would mean that the block holds a QC with a view < 1, which should never happen. Rather than have an edge case to handle this, but just side-step it entirely by guaranteeing that blocks in production all have view > 0. When the PoW -> PoS cutover happens, the first transition block will have its view = blockHeight